046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,86 @@
+```
+status:
+    <kafka_pod_name>:
+      restartTimeStamp: 
+      logRecoveryExceededTimeout: <true|false>
+```
+- When KafkaRoller restarts a broker, write the creation timestamp to the new field, `restartTimeStamp`.
+- If a kafka pod does not become ready within the operational timeout, check the broker state by sending a request to the new endpoint:
+    - If log recovery is in progress, apply the new logic for handling log recovery.
+    - Otherwise, continue with the current behaviour.
+
+*Logic for handling log recovery*
+- Wait for log recovery to complete until the operation timeout (5 mins by default) reaches.
+- While waiting for log recovery to complete, check its progress by periodically sending a request to the new endpoint for `remainingLogsToRecover` and `remainingSegmentsToRecover` metrics and log them in an `INFO` message.
+- If the operation timeout is exceeded, check the value of `logRecoveryExceededTimeout` field in the CR:
+    - If false, log a `WARN` message, fail the reconciliation and set `logRecoveryExceededTimeout` to true.
+    - If true, fail the reconciliation.
+- If kafka pod becomes ready, check the value of `logRecoveryExceededTimeout`. If set to true:
+    - Calculate the time passed since the `restartTimeStamp` and log it in a WARN message. This would be helpful to figure out a more appropriate timeout value to avoid this in the future.
+    - Empty `restartTimeStamp` and `logRecoveryExceededTimeout` values.
+
+### For future consideration:
+
+- The operator might need access to more broker metrics. We can easily implement additional endpoints to expose them.
+- The readiness check could also make use of the KafkaAgent endpoint to get the broker state instead of checking for a file on disk. An internal non-TLS endpoint would need to be implemented for it which can be elaborated on in a further proposal.
+- The liveness and readiness checks potentially will need to access non-yammer metrics for KRaft, which could be exposed in the same manner, but would need a new mechanism to collect them.
+
+## Affected projects
+
+* strimzi/kafka-cluster-operator
+
+## Compatibility
+
+After the operator gets updated with the KafkaRoller change, it will expect the KafkaAgent endpoint to be present. However, the KafkaAgent may not have been updated yet.
+
+To maintain backwards compatibility, use the new logic that handles log recovery only if KafkaRoller is able to successfully connect to the KafkaAgent endpoint. If the KafkaAgent endpoint is not available yet or the agent is not running, the current logic of restarting the broker after a timeout period should be used. This ensures compatibility with previous versions. Additionally, in the case that the KafkaAgent API is available but the agent is stuck or not running for some reason, the current logic's force restart of the broker may help recover the agent.
```

Comment:
I think this is missing more context of the whole roller. Here (but basically also in the previous section) you describe some steps. But it is not clear where exactly they fit into the KafkaRoller. I think for general understanding it might be best to describe the whole Roller process and how it works today and explain which parts you change and how.

As an example ... you suggest here that if it cannot reach the Agent API, it will be force-rolled. But that is certainly not the right approach for upgrades. In the upgrades, regardless of whether you reach the Agent API, if the pod is ready, you do not force roll it but you need to connect to it with Kafka Admin API and roll it through the regular process which follows the in-sync replicas etc.

It is not really clear if that is intended that way or not since it is not clear which part of the roller you are proposing to change.

Reply:
The intention is not to have an additional logic there to force roll the broker if it can't reach the API. It would just follow the current behaviour which waits for broker to become ready and may force roll if it times out. I was trying to say, following the current behaviour is ok when the API is not reachable. If the agent is stuck or  not running, that would likely to be the action to take to get it working anyway. 

I will try make this clearer and how it fits to the KafkaRoller logic.